src/dvc_data/build.py

```diff
@@ -76,6 +78,74 @@ def _build_file(path, fs, name, odb=None, upload_odb=None, dry_run=False):
     return meta, obj
 
 
+def _find_files(path, fs, ignore=None):
+    if ignore:
+        walk_iter = ignore.walk(fs, path)
+    else:
+        walk_iter = fs.walk(path)
+
+    for root, _, fnames in walk_iter:
+        if DefaultIgnoreFile in fnames:
+            raise IgnoreInCollectedDirError(
+                DefaultIgnoreFile, fs.path.join(root, DefaultIgnoreFile)
+            )
+
+        # NOTE: we know for sure that root starts with path, so we can use
+        # faster string manipulation instead of a more robust relparts()
+        rel_key: Tuple[Optional[Any], ...] = ()
+        if root != path:
+            rel_key = tuple(root[len(path) + 1 :].split(fs.sep))
+
+        # NOTE: empty `fname` might happen with s3/gs/azure/etc, where empty
+        # objects like `dir/` might be used to create an empty dir
+        yield from (
+            (root + fs.sep + fname, (*rel_key, fname))
+            for fname in fnames
+            if fname
+        )
+
+
+def _diff_state_to_dir(path, fs, ignore=None, state=None, pbar=None):
+    files = dict(_find_files(path, fs, ignore=ignore))
```

Comment:
This is us scanning for files, which might take a bit of time, so we will eventually (or even now) add a spinner or something for it. So lets move this out (e.g. to right before `Tqdm` in `_build_tree`) and then pass the result here.

Reply:
Should reduce the complication with `pbar.refresh` and `ignore` too, btw.